mainapp/src/Admin.tsx

```diff
@@ -1,30 +1,389 @@
+
+            </Stack>
+        </Box>
+    </Modal >;
+}
+
+function deleteUser(users: User[], user: User, updateModel: React.Dispatch<Msg>, setModal: React.Dispatch<React.SetStateAction<boolean>> = () => console.log("modal closed")) {
+    const updatedUsers = users.filter(prevUsers => prevUsers.key !== user.key);
+    console.log(`user ${user.key} deleted`)
+    console.log(updatedUsers)
+    return () => {
+        putUsers('/users', updatedUsers)
+            .then((users) => updateModel({ type: 'ServerRespondedWithUsers', payload: success(users) }))
+            .then(() => setModal(false))
+            .catch((err) => updateModel({ type: 'ServerRespondedWithUsers', payload: failure(err.msg) }));
+    };
+}
+
+function saveUser(updateModel: React.Dispatch<Msg>, users: User[], user: User, localUser: User, setModal: React.Dispatch<React.SetStateAction<boolean>>) {
+    return () => {
+
+        updateModel({ type: 'UserClickedSaveChanges', payload: {} });
+        putUsers('/users', users.map(u => u.key === user.key ? localUser : u))
+            .then((person) => updateModel({ type: 'ServerRespondedWithUsers', payload: success(person) }))
+            .then(() => setModal(false))
+            .catch((err) => updateModel({ type: 'ServerRespondedWithUsers', payload: failure(err.msg) }));
+    };
+}
+
+function restoreUsers(updateModel: React.Dispatch<Msg>, setModal: React.Dispatch<React.SetStateAction<boolean>>) {
+    console.log("restore user")
+    return () => {
+
+        getUsers('/users')
+            .then((person) => updateModel({ type: 'ServerRespondedWithUsers', payload: success(person) }))
+            .then(() => setModal(false))
+            .catch((err) => updateModel({ type: 'ServerRespondedWithUsers', payload: failure(err.msg) }));
+    };
+}
+
+
+
+type UsersProps = {
+    users: User[],
+}
+
+const Users: React.FC<UsersProps> = ({ users }): JSX.Element => {
+    const { model, updateModel } = useModel();
+    return (
+        <Box sx={{ justifyContent: 'center', display: 'flex' }}>
+            <TableContainer component={Paper}>
+                <Table sx={{ width: '100%' }}>
+                    <TableHead>
+                        <TableRow >
+                            <TableCell>Login</TableCell>
+                            <TableCell>Groups</TableCell>
+                            <TableCell>Actions</TableCell>
+                        </TableRow>
+                    </TableHead>
+                    <TableBody>{users.map(user => (<TableRow key={user.key}>
+                        <TableCell>
+                            {user.login}
+                        </TableCell>
+                        <TableCell>
+                            {user.groups.join(', ')}
+                        </TableCell>
+                        <TableCell>
+                            <ButtonGroup>
+                                <User key={user.key} user={user}></User>
```

Comment:
The user.key is only used here, maybe we can find another way to define a unique key here and not store it into the json file if it is not needed. What do you think?